ta/pkcs11/src/processing_ec.c

```diff
@@ -0,0 +1,648 @@
+{
+	size_t idx = 0;
+
+	/* Weak: not a real DER parser: try by params then by named curve */
+	for (idx = 0; idx < ARRAY_SIZE(ec_curve_param); idx++) {
+		const struct supported_ecc_curve *curve = &ec_curve_param[idx];
+
+		if (size == curve->oid_size &&
+		    !TEE_MemCompare(attr, curve->oid_der, curve->oid_size))
+			return curve;
+
+		if (size == curve->name_size &&
+		    !TEE_MemCompare(attr, curve->name_der, curve->name_size))
+			return curve;
+	}
+
+	return NULL;
+}
+
+size_t ec_params2tee_keysize(void *ec_params, size_t size)
+{
+	const struct supported_ecc_curve *curve = get_curve(ec_params, size);
+
+	if (!curve)
+		return 0;
+
+	return curve->key_size;
+}
+
+/*
+ * This function intentionally panics if the curve is not found.
+ * Use ec_params2tee_keysize() to check the curve is supported by
+ * the internal core API.
+ */
+uint32_t ec_params2tee_curve(void *ec_params, size_t size)
+{
+	const struct supported_ecc_curve *curve = get_curve(ec_params, size);
+
+	assert(curve);
+
+	return curve->tee_id;
+}
+
+uint32_t load_tee_ec_key_attrs(TEE_Attribute **tee_attrs, size_t *tee_count,
+			       struct pkcs11_object *obj)
+{
+	TEE_Attribute *attrs = NULL;
+	size_t count = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
+	assert(get_key_type(obj->attributes) == PKCS11_CKK_EC);
+
+	switch (get_class(obj->attributes)) {
+	case PKCS11_CKO_PUBLIC_KEY:
+		attrs = TEE_Malloc(3 * sizeof(TEE_Attribute),
+				   TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!attrs)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_ECC_CURVE,
+				       obj, PKCS11_CKA_EC_PARAMS))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_ECC_PUBLIC_VALUE_X,
+				       obj, PKCS11_CKA_EC_POINT))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_ECC_PUBLIC_VALUE_Y,
+				       obj, PKCS11_CKA_EC_POINT))
+			count++;
+
+		if (count == 3)
+			rc = PKCS11_CKR_OK;
+
+		break;
+
+	case PKCS11_CKO_PRIVATE_KEY:
+		attrs = TEE_Malloc(4 * sizeof(TEE_Attribute),
+				   TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!attrs)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		if (pkcs2tee_load_attr(&attrs[count], TEE_ATTR_ECC_CURVE,
+				       obj, PKCS11_CKA_EC_PARAMS))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_ECC_PRIVATE_VALUE,
+				       obj, PKCS11_CKA_VALUE))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_ECC_PUBLIC_VALUE_X,
+				       obj, PKCS11_CKA_EC_POINT))
+			count++;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+				       TEE_ATTR_ECC_PUBLIC_VALUE_Y,
+				       obj, PKCS11_CKA_EC_POINT))
+			count++;
+
+		if (count == 4)
+			rc = PKCS11_CKR_OK;
+
+		break;
+
+	default:
+		assert(0);
+		break;
+	}
+
+	if (rc == PKCS11_CKR_OK) {
+		*tee_attrs = attrs;
+		*tee_count = count;
+	} else {
+		TEE_Free(attrs);
+	}
+
+	return rc;
+}
+
+uint32_t pkcs2tee_algo_ecdsa(uint32_t *tee_id,
```

Comment:
s/uint32_t/enum pkcs11_rc/

few more occurrences i think.

Reply:
Right -- I scan for them.